File: solutions/ocaml/custom-set/1/custom_set.ml
(* Dedicated to Junko F. Didi and Shree DR.MDD *)

module type ELEMENT = sig
  type t
  val compare : t -> t -> int
end

module Make(El: ELEMENT) = struct
  type t = El.t list
  type el = El.t

  let el_equal a b = El.compare a b = 0

  let is_empty lst = match lst with [] -> true | _ -> false

  let is_member s x = List.exists (fun y -> El.compare x y = 0) s

  let is_subset s1 s2 = List.for_all (is_member s2) s1

  let is_disjoint s1 s2 = List.for_all (is_member s2 |> Fun.negate) s1

  let equal s1 s2 = is_subset s1 s2 && is_subset s2 s1

  let of_list xs = xs

  let add s x = x :: s

  type status = [
    | `OnlyA
    | `OnlyB
    | `Both
  ]

  let diff_filter _ _ _ = []

  let difference s1 s2 = List.filter (is_member s2 |> Fun.negate) s1

  let intersect s1 s2 = List.filter (is_member s2) s1

  let union s1 s2 = s1 @ s2
end

File: solutions/go/custom-set/1/custom_set.go
// Dedicated to Shree DR.MDD
package stringset

import (
	"sort"
	"strconv"
	"strings"
)

type Set map[string]struct{}

func New() Set {
	return make(Set)
}

func NewFromSlice(elements []string) Set {
	container := New()
	for _, item := range elements {
		container.Add(item)
	}
	return container
}

func (s Set) String() string {
	if len(s) == 0 {
		return "{}"
	}
	list := make([]string, 0, len(s))
	for value := range s {
		list = append(list, strconv.Quote(value))
	}
	sort.Strings(list)
	return "{" + strings.Join(list, ", ") + "}"
}

func (s Set) IsEmpty() bool {
	return len(s) == 0
}

func (s Set) Has(elem string) bool {
	_, ok := s[elem]
	return ok
}

func (s Set) Add(elem string) {
	s[elem] = struct{}{}
}

func Subset(a, b Set) bool {
	for value := range a {
		if !b.Has(value) {
			return false
		}
	}
	return true
}

func Disjoint(a, b Set) bool {
	for value := range a {
		if b.Has(value) {
			return false
		}
	}
	return true
}

func Equal(a, b Set) bool {
	if len(a) != len(b) {
		return false
	}
	return Subset(a, b)
}

func Intersection(a, b Set) Set {
	result := New()
	for value := range a {
		if b.Has(value) {
			result.Add(value)
		}
	}
	return result
}

func Difference(a, b Set) Set {
	result := New()
	for value := range a {
		if !b.Has(value) {
			result.Add(value)
		}
	}
	return result
}

func Union(a, b Set) Set {
	result := New()
	for value := range a {
		result.Add(value)
	}
	for value := range b {
		result.Add(value)
	}
	return result
}
